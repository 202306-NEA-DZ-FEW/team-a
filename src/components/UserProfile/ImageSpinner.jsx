import Image from "next/image";
import Spinner from "public/images/spinner.svg";

function ImageSpinner({ classes }) {
  return (
    <div className={classes}>
      <Image
        className='opacity-60 w-1/2 h-1/2 object-contain'
        src={Spinner}
        alt='loading'
        height={32}
        width={32}
        priority
      />
    </div>
  );
}

export default ImageSpinner;
