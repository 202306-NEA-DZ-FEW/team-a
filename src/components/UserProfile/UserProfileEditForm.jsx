import { doc, updateDoc } from "firebase/firestore";
import { useFormik } from "formik";
import { useTranslation } from "next-i18next";
import { toast } from "react-toastify";
import * as Yup from "yup";

import { db } from "@/lib/firebase";
import getAllStates from "@/lib/getAllStates";

import Input from "../Input";
import SelectInput from "../SelectInput";

function UserProfileEditForm({ userData, onUpdate }) {
  const { t } = useTranslation();
  const states = getAllStates(t);

  const formik = useFormik({
    initialValues: {
      name: userData.name,
      email: userData.email,
      phone: userData.phone,
      location: userData.location,
    },
    validationSchema: Yup.object({
      name: Yup.string()
        .max(20, t("dashboard:userUpadteForm:nameLength"))
        .required(t("dashboard:userUpadteForm:nameRequired")),
      email: Yup.string()
        .email(t("dashboard:userUpadteForm:emailValid"))
        .required(t("dashboard:userUpadteForm:emailRequired")),
      phone: Yup.number().required(t("dashboard:userUpadteForm:phoneRequired")),
      location: Yup.string().required(
        t("dashboard:userUpadteForm:locationRequired")
      ),
    }),
    onSubmit: async (values) => {
      const userDocRef = doc(db, "users", userData.uid);
      try {
        await updateDoc(userDocRef, values);
        onUpdate({
          ...values,
          photoURL: userData.photoURL,
          uid: userData.uid,
          role: userData.role,
          date: userData.date,
        });
        toast.success("Profile updated successfully âœ…", {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
      } catch (error) {
        toast.error(error.code, {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 1500,
        });
      } finally {
        document.getElementById("my_modal_1").close();
      }
    },
  });

  return (
    <dialog id='my_modal_1' className='modal bg-black bg-opacity-40'>
      <div className='modal-box bg-white transition-all duration-500 ease-in-out'>
        <form onSubmit={formik.handleSubmit} className='flex flex-col gap-4'>
          <h1 className='text-xl font-bold text-center'>
            Update your personal profile
          </h1>
          <Input
            name='name'
            type='text'
            placeholder={t("dashboard:userUpadteForm:namePlaceholder")}
            label={t("dashboard:userUpadteForm:nameLabel")}
            handleChange={formik.handleChange}
            value={formik.values.name}
            handleBlur={formik.handleBlur}
            error={formik.errors.name}
            touched={formik.touched.name}
          />
          <Input
            name='email'
            type='email'
            placeholder={t("dashboard:userUpadteForm:emailPlaceholder")}
            label={t("dashboard:userUpadteForm:emailLabel")}
            handleChange={formik.handleChange}
            value={formik.values.email}
            handleBlur={formik.handleBlur}
            error={formik.errors.email}
            touched={formik.touched.email}
          />
          <Input
            name='phone'
            type='text'
            placeholder={t("dashboard:userUpadteForm:phonePlaceholder")}
            label={t("dashboard:userUpadteForm:phoneLabel")}
            handleChange={formik.handleChange}
            value={formik.values.phone}
            handleBlur={formik.handleBlur}
            error={formik.errors.phone}
            touched={formik.touched.phone}
          />
          <SelectInput
            name='location'
            data={states}
            label={t("dashboard:userUpadteForm:locationLabel")}
            handleChange={formik.handleChange}
            value={formik.values.location}
            touched={formik.touched.location}
            error={formik.errors.location}
          />
          <div className='modal-action'>
            <button
              type='submit'
              className='btn rounded-full normal-case tracking-wider btn-active'
            >
              {t("dashboard:userUpadteForm:submit")}
            </button>
            <div method='dialog'>
              <button
                type='button'
                className='btn btn-outline rounded-full normal-case tracking-wider font-light'
                onClick={() => document.getElementById("my_modal_1").close()}
              >
                {t("dashboard:userUpadteForm:cancel")}
              </button>
            </div>
          </div>
        </form>
      </div>
    </dialog>
  );
}

export default UserProfileEditForm;
