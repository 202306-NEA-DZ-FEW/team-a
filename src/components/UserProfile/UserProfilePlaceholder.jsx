function UserProfilePlaceholder() {
  return (
    <div className='flex flex-col w-full md:flex-row gap-8 justify-center md:items-start items-center'>
      <section className='flex flex-col gap-8 md:gap-1 items-center justify-between w-full md:flex-row animate-pulse'>
        <div className='md:w-[120px] w-full self-start flex md:flex-col flex-row gap-2'>
          <div className='w-[100px] self-start h-[100px] rounded-full bg-gray-300 animate-pulse'></div>
          <div>
            <div className='w-[100px] my-2 self-start h-[10px] rounded-full bg-gray-300 animate-pulse'></div>
            <div className='w-[80px] self-start h-[10px] rounded-full bg-gray-300 animate-pulse'></div>
          </div>
        </div>
        <div className='w-[80%] h-[110px] rounded-xl bg-gray-300 animate-pulse'></div>
      </section>
    </div>
  );
}

export default UserProfilePlaceholder;
