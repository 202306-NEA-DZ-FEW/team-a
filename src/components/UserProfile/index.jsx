import Image from "next/image";
import { useTranslation } from "next-i18next";
import profile from "public/images/profile.svg";
import { useState } from "react";
import { BsFillPencilFill, BsImageFill } from "react-icons/bs";
import { MdOutlineShareLocation } from "react-icons/md";

import { getLocationName } from "@/lib/helpers";
import useImageUpload from "@/lib/useImageUpload";

import { useAuth } from "@/context/AuthProvider";

import EditProfileForm from "./EditProfileForm";
import ImageSpinner from "./ImageSpinner";

function UserProfile({ userData }) {
  const [profileData, setProfileData] = useState(userData);
  const { setUser } = useAuth();
  const { i18n, t } = useTranslation();
  const loaction = t(`states:${profileData.location}`);
  const locationName = getLocationName(loaction);
  const [selectedImage, setSelectedImage] = useState(null);
  const { loading, updateImage } = useImageUpload();

  const handleImageChange = async (e) => {
    const imageFile = e.target.files[0];
    const downloadURL = await updateImage(profileData, imageFile);
    setSelectedImage(downloadURL);
    setUser({ ...userData, photoURL: downloadURL });
  };

  const handleUpdateProfile = (newUserData) => {
    setProfileData(newUserData);
  };

  return (
    <div className='flex flex-col md:flex-row gap-4 justify-center items-center'>
      <section className='flex rounded-2xl drop-shadow-md flex-row gap-4 md:gap-1 items-center justify-center w-full md:w-1/4 md:flex-col'>
        <div className='avatar m-[6px]'>
          <div className='w-20 rounded-full relative ring ring-primary ring-offset-base-100 ring-offset-2 group'>
            {loading && (
              <ImageSpinner classes='absolute w-full h-full p-5 rounded-full bg-white bg-opacity-30' />
            )}
            <Image
              className='object-cover w-full h-full'
              height={150}
              width={150}
              priority
              src={selectedImage || profileData.photoURL || profile}
              alt='User Profile'
            />
            <div
              type='file'
              className='absolute hidden w-full h-full bg-white rounded-full top-0 left-0 group-hover:flex cursor-pointer bg-opacity-50 '
            >
              <input
                type='file'
                id='fileInput'
                onChange={handleImageChange}
                accept='image/*'
                className='absolute hidden text-xs'
              />
              <label
                htmlFor='fileInput'
                className='text-4xl w-full h-full flex items-center justify-center cursor-pointer'
              >
                <BsImageFill size={24} />
              </label>
            </div>
          </div>
        </div>
        <div className='md:text-center text:start w-full'>
          <h3 className='font-bold text-xl'>{profileData.name}</h3>
          <p
            className='text-md flex items-center justify-center gap-1 text-gray-600'
            dir={i18n?.language == "ar" ? "rtl" : "ltr"}
          >
            <span>
              <MdOutlineShareLocation />
            </span>
            {locationName}
          </p>
        </div>
      </section>
      {/* user info............................... */}
      <section
        className='flex md:justify-between md:items-center tracking-wider md:flex-row gap-2 bg-white drop-shadow-md md:gap-6 w-full rounded-2xl'
        dir={i18n?.language == "ar" ? "rtl" : "ltr"}
      >
        <div className='flex flex-col md:flex-row gap-2 p-6 md:gap-6 flex-1'>
          <div className='flex flex-col gap-2'>
            <p>
              <span className='font-bold'>
                {t("dashboard:userinfo:name")}:{" "}
              </span>
              <span>{profileData.name}</span>
            </p>
            <p>
              <span className='font-bold'>
                {t("dashboard:userinfo:location")}:{" "}
              </span>
              <span>{locationName}</span>
            </p>
          </div>
          <div className='flex flex-col gap-2'>
            <p>
              <span className='font-bold'>
                {t("dashboard:userinfo:email")}:{" "}
              </span>
              <span>{profileData.email}</span>
            </p>
            <p>
              <span className='font-bold'>
                {t("dashboard:userinfo:phoneNumber")}:{" "}
              </span>
              <span>{profileData.phone ? profileData.phone : "-"}</span>
            </p>
          </div>
          <p className='self-start'>
            <span className='font-bold'>
              {t("dashboard:userinfo:language")}:{" "}
            </span>
            <span>{i18n.language}</span>
          </p>
        </div>
        <div
          className={`self-start md:self-stretch md:bg-gray-100 ${
            i18n?.language == "ar"
              ? "md:rounded-tl-2xl md:rounded-bl-2xl"
              : "md:rounded-tr-2xl md:rounded-br-2xl"
          }`}
        >
          <button
            onClick={() => document.getElementById("my_modal_1").showModal()}
            className={`btn transition-all duration-500 ease-in-out group btn-ghost hover:bg-opacity-10 rounded-2xl h-full
            ${
              i18n?.language == "ar"
                ? "md:rounded-tr-none md:rounded-br-none"
                : "md:rounded-tl-none md:rounded-bl-none"
            }
            `}
          >
            <BsFillPencilFill className='text-gray-400 text-xl md:mx-2 transition-all duration-500 ease-in-out group-hover:text-gray-800' />
          </button>
          <EditProfileForm
            userData={profileData}
            onUpdate={handleUpdateProfile}
          />
        </div>
      </section>
    </div>
  );
}

export default UserProfile;
