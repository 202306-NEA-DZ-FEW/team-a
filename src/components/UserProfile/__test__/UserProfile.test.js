import renderer from "react-test-renderer";

import UserProfile from "..";

jest.mock("@/lib/firebase", () => {
  return {
    app: {}, // Mock app object
    db: {}, // Mock db object
    storage: {}, // Mock storage object
    auth: {}, // Mock auth object
  };
});

it("renders correctly", () => {
  const mockUser = {
    name: "John",
    email: "jhm@gmail.com",
    phone: "123",
    location: "1- Adrad",
  };
  const tree = renderer.create(<UserProfile userData={mockUser} />).toJSON();
  expect(tree).toMatchSnapshot();
});
