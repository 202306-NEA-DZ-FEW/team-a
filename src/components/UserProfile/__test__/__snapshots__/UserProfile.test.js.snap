// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex flex-col md:flex-row gap-4 justify-center items-start"
>
  <section
    className="flex flex-row gap-2 items-center justify-center w-full md:w-1/4 md:flex-col"
  >
    <figure
      className="relative group"
    >
      <img
        alt="User Profile"
        className="object-cover md:w-32 w-28 rounded-full"
        data-nimg="1"
        decoding="async"
        height={150}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
        style={
          Object {
            "color": "transparent",
          }
        }
        width={150}
      />
      <div
        className="absolute hidden text-center bg-white rounded-full h-full md:w-32 w-28 top-0 group-hover:flex cursor-pointer bg-opacity-40 items-center justify-center"
        type="file"
      >
        <input
          accept="image/*"
          className="absolute hidden text-xs"
          id="fileInput"
          onChange={[Function]}
          type="file"
        />
        <label
          className="text-4xl cursor-pointer"
          htmlFor="fileInput"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.002 3a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-12a2 2 0 0 1-2-2V3zm1 9v1a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V9.5l-3.777-1.947a.5.5 0 0 0-.577.093l-3.71 3.71-2.66-1.772a.5.5 0 0 0-.63.062L1.002 12zm5-6.5a1.5 1.5 0 1 0-3 0 1.5 1.5 0 0 0 3 0z"
            />
          </svg>
        </label>
      </div>
    </figure>
    <div
      className="md:text-center text:start w-full"
    >
      <h3
        className="font-bold text-xl"
      >
        John
      </h3>
      <p
        className="text-md text-gray-600"
      >
        states:1- Adrad
      </p>
    </div>
  </section>
  <section
    className="flex md:justify-between md:items-center md:flex-row mt-6 gap-2 bg-gray-200 md:gap-6 w-full rounded-xl"
    dir="ltr"
  >
    <div
      className="flex flex-col md:flex-row gap-2 p-6 md:gap-6 flex-1"
    >
      <div
        className="flex flex-col gap-2"
      >
        <p>
          <span
            className="font-bold"
          >
            userinfo.Name
            :
             
          </span>
          <span>
            John
          </span>
        </p>
        <p>
          <span
            className="font-bold"
          >
            userinfo.Location
            :
             
          </span>
          <span>
            states:1- Adrad
          </span>
        </p>
      </div>
      <div
        className="flex flex-col gap-2"
      >
        <p>
          <span
            className="font-bold"
          >
            userinfo.Email
            :
             
          </span>
          <span>
            jhm@gmail.com
          </span>
        </p>
        <p>
          <span
            className="font-bold"
          >
            userinfo.Phonenumber
            :
             
          </span>
          <span>
            123
          </span>
        </p>
      </div>
      <p
        className="self-start"
      >
        <span
          className="font-bold"
        >
          userinfo.language
          :
           
        </span>
        <span>
          en
        </span>
      </p>
    </div>
    <div
      className="self-start md:self-stretch md:bg-gray-300 md:rounded-tr-xl md:rounded-br-xl"
    >
      <button
        className="btn btn-ghost h-full"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.854.146a.5.5 0 0 0-.707 0L10.5 1.793 14.207 5.5l1.647-1.646a.5.5 0 0 0 0-.708l-3-3zm.646 6.061L9.793 2.5 3.293 9H3.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.207l6.5-6.5zm-7.468 7.468A.5.5 0 0 1 6 13.5V13h-.5a.5.5 0 0 1-.5-.5V12h-.5a.5.5 0 0 1-.5-.5V11h-.5a.5.5 0 0 1-.5-.5V10h-.5a.499.499 0 0 1-.175-.032l-.179.178a.5.5 0 0 0-.11.168l-2 5a.5.5 0 0 0 .65.65l5-2a.5.5 0 0 0 .168-.11l.178-.178z"
          />
        </svg>
      </button>
    </div>
  </section>
</div>
`;
